spec/consensus/gossip/gossip.md

```diff
@@ -0,0 +1,333 @@
+# Part 1: Background
+
+> **Warning**
+> We assume that you understand the Tendermint algorithm and therefore we will not review it here.
+If this is not the case, please refer to [here](../).
+
+Three kinds of messages are exchanged in the Tendermint algorithm: `PROPOSAL`, `PRE-VOTE`, and `PRE-COMMIT`.
+The algorithm progresses when certain conditions are satisfied over the set of messages received.
+For example, in order do decide on a value `v`, the set must include a `PROPOSAL` for `v` and more than two thirds of the number processes in `PRE-COMMIT` for the same `v`, during the same round.
+
+To ensure progress and termination, the algorithm assumes a **Global Stabilization Time (GST)**, after which communication is reliable and timely (Eventual $\Delta$-Timely Communication), which is used to provide **Gossip Communication**.
+
+| Eventual $\Delta$-Timely Communication|
+|-----|
+|There is a bound $\Delta$ and an instant GST (Global Stabilization Time) such that if a correct process $p$ sends a message $m$ at a time $t \geq \text{GST}$ to a correct process $q$, then $q$ will receive $m$ before $t + \Delta$.
+
+|Gossip communication|
+|-----|
+| (i) If a correct process $p$ sends some message $m$ at time $t$, all correct processes will receive $m$ before $\text{max} \{t,\text{GST}\} + \Delta$.
+| (ii) If a correct process $p$ receives some message $m$ at time $t$, all correct processes will receive $m$ before $\text{max}\{t,\text{GST}\} + \Delta$.
+
+Roughly speaking, if Gossip Communication is guaranteed, then all messages sent by correct processes will be eventually delivered to all correct processes and allow all correct processes to realize that the same conditions are.
+
+However, since processes are subject to failures, it is not guaranteed that proposals are ever sent and correct processes cannot wait indefinitely.
+Hence, processes execute in rounds in which they wait for conditions to be met for sometime and, if they timeout, send negative messages that will lead to new rounds.
+Again, Gossip Communication guarantees that eventually the conditions for deciding are met, even if only after GST is reached.
+
+Implementing Gossip Communication, however, is hard because even messages sent before the GST need to be buffered to be reliably delivered between correct processes, but since GST may take indefinitely long to arrive, in practice implementing this property would require unbounded memory.
+
+Fortunately, while Gossip Communication is a sufficient condition for the Tendermint algorithm to terminate, it is not strictly necessary, because the conditions to progress and terminate are evaluated over the messages of subsets of rounds executed, not all of them, and as new rounds are executed, messages in old rounds may be become obsolete and be ignored and forgotten, saving memory.
+In other words, the algorithm does not require all messages to be delivered, only messages that advance the state of the processes.
```

Comment:
No real-world implementation of a distributed system over a geographically diverse network can satisfy the definition of GST which asserts that "if a correct process P sends a message at time T to a correct process Q, then Q will receive that message before T+∆".

Nothing can guarantee the delivery of a message from P to Q. You can say that delivery failure to Q marks Q as "incorrect", but that means "correctness" is necessarily a property that's relative to a specific observer. Q can be incorrect from P1, but correct from P2. There is no single global truth about Q's correctness.

Weakening Gossip Communication to be "not strictly necessary" because "the algorithm [only requires] messages that advance the state of the processes" doesn't really solve anything. Those requisite messages are still (presumably) delivered over gossip channels, right? If so, then they still require reliable delivery; in fact delivery guarantees at an e.g. gossip layer are not really different based on properties of any individual message. 

Gossip communication is not guaranteed, and is not reliable, just by definition. So you can't assume reliable delivery of gossiped messages at a higher layer e.g. CometBFT. And that means you can't rely on any specific guarantees of higher-layer concepts like GST in your application code.

Reply:
It is impossible to guarantee progress of consensus in asynchronous distributed systems in which failures can happen. Period.

GST is used as tool to prove that if there are stable periods, the algorithm will progress, the separate algorithms from others that wouldn't progress even if there were stable periods.

No one expects GST to happen in practice. Still, implementations of Tendermint, Paxos, Raft and many others have been used for years, and they keep making progress during stable periods. If those stable periods cease, all these systems will stop working.

In conclusion, your WEB3 application working is a gift from the stable periods. You can rely on them.

Reply:
> It is impossible to guarantee progress of consensus in asynchronous distributed systems in which failures can happen. Period.

Definitely! This is also established by e.g. CAP. You pick either availability (which weakens your consistency model if there is a partition) or consistency (which halts your system if there is a partition). You can have progress, or consistency, but not both. No problem! Systems like Paxos and Raft and I guess Tendermint pick consistency. But that means, in the presence of a partition, the system will necessarily halt. It will pick up again when the partition is healed, but how long that takes is unknowable. Concepts like stability periods or a GST are I guess different ways to characterize partition-free windows of time in a system, but those concepts are emergent and descriptive properties of that system, not guarantees, and actually unknowable at any individual node. Protocols execute at individual nodes and without whole-system knowledge, so those system-level concepts can't be part of a protocol definition.

This is understood by e.g. Raft and Paxos, neither of which define or rely on any notion of a stability period or a GST, or assume anything about stability in the guarantees they provide to their consumers.

In contrast, this spec asserts stuff like

> To ensure progress and termination, the algorithm assumes a GST after which communication is reliable and timely

which isn't really, er, possible. Certainly neither Raft nor Paxos provide this assurance. In fact no system can ensure both consistency and progress/termination. And a distributed system protocol cannot rely on an assumption that delivery will be reliable or timely within some time boundary. If your system model assumes a GST as specified here, it only means that the model becomes invalid when that assumption fails. Which it will.

Reply:
Let me repeat myself.

> GST is used as a tool to prove that if there are stable periods, the algorithm will progress, to separate algorithms from others that wouldn't progress even if there were stable periods.
> No one expects GST to happen in practice. Still, implementations of Tendermint, Paxos, Raft and many others have been used for years, and they keep making progress during stable periods. If those stable periods cease, all these systems will stop working.

Again, it is a tool to prove that under the right conditions, the algorithm progresses and If these conditions do not happen. Why do we do it then? To differentiate these algorithms (which Guerraoui et al. called indulgent) from algorithms that are safe but don't progress even under the right conditions (which are technically correct but useless). So we need to make these assumptions here and show that when they hold, the algorithm progresses.

WRT your statement

> This is understood by e.g. Raft and Paxos, neither of which define or rely on any notion of a stability period or a GST, or assume anything about stability in the guarantees they provide to their consumers.

Both Paxos and Raft make timing assumptions. For [example](https://lamport.azurewebsites.net/pubs/lamport-paxos.pdf)

> Progress could not be guaranteed unless enough legislators stayed in the Cham- ber for a long enough time. Because Paxon legislators were unwilling to curtail their outside activities, it was impossible to ensure that any decree would ever be passed. However, legislators were willing to guarantee that, while in the Chamber, they and their aides would act promptly on all parliamentary matters. This guar- antee allowed the Paxons to devise a parliamentary protocol satisfying the following progress condition.
> > If a majority of the legislators2 were in the Chamber and no one entered or left the Chamber for a sufficiently long period of time then any decree proposed by a legislator in the Chamber would be passed, and every decree that had been passed would appear in the ledger of every legislator in the Chamber.

and 

> Achieving the progress condition required that legislators be able to measure the passage of time, so they were given simple hourglass timers.

The fact that timing assumptions are often hidden behind Omega oracles, which ensure eventual stable leader election, makes it harder to perceive them, but they are there.

Reply:
Does a Paxos node know itself whether or not it is a valid legislator, or that it is in the Chamber, for a given round?